src/lib.rs

```diff
@@ -202,21 +269,36 @@ impl PerformanceData {
             SetTimeFlags::Default,
         );
         while current <= end {
+            aperf_collect_data.time = TimeEnum::DateTime(Utc::now());
+            aperf_collect_data.data = HashMap::new();
             let ret = tfd.read();
             if ret > 1 {
                 error!("Missed {} interval(s)", ret - 1);
             }
             debug!("Time elapsed: {:?}", start.elapsed());
             current += time::Duration::from_secs(ret * self.init_params.interval);
-            for (_name, datatype) in self.collectors.iter_mut() {
+            for (name, datatype) in self.collectors.iter_mut() {
                 if datatype.is_static {
                     continue;
                 }
-                datatype.collect_data()?;
-                datatype.write_to_file()?;
+                let collect_time: u64 = datatype.collect_data()? as u64;
+                let print_time: u64 = datatype.write_to_file()? as u64;
+                aperf_collect_data
+                    .data
+                    .insert(name.clone() + "-collect", collect_time);
+                aperf_collect_data
+                    .data
+                    .insert(name.clone() + "-print", print_time);
```

Comment:
Encapsulating the measurements like this can be made more robust by taking advantage of passing a function to a helper
```
aperf_collect_data.measure(name.clone() + "-collect", |datatype| datatype.collect_data())```
```
I'm sure there's syntax errors in there, but it's just an example

such a ```measure``` function can do the timing and the appending of the data to ```self``` (```aperf_collect_data```). It removes the need for repeated timing code and avoids the risk of, for example in this case, of the collect and print times getting mixed up. It's also less code in every place doing this.

Reply:
Yeah. Changed it to have a measure() which does the timing and updates the self.data.